datafusion/optimizer/src/type_coercion.rs

```diff
@@ -164,11 +163,61 @@ impl ExprRewriter for TypeCoercionRewriter<'_> {
                 };
                 expr.rewrite(&mut self.const_evaluator)
             }
+            Expr::InList {
+                expr,
+                list,
+                negated,
+            } => {
+                let expr_data_type = expr.get_type(&self.schema)?;
+                let list_data_types = list
+                    .iter()
+                    .map(|list_expr| list_expr.get_type(&self.schema))
+                    .collect::<Result<Vec<_>>>()?;
+                let result_type =
```

Comment:
Do we need a common function in a new file like `in_list_rule.rs` to do the `coerce_types`?
cc @alamb @andygrove

Reply:
it is also can be used in the `in_list.rs` for physical inlist test cases.

Reply:
I will a issue to track this and centralize the code together for the inlist